### PR TITLE
Fix nightly build failures

### PR DESCRIPTION
**Pull request by neilmayhew**
# Description

* Fix nightly build failures
* Allow simulating a nightly build when manually triggering a build

# Checklist

- [x] Commit sequence broadly makes sense and commits have useful messages
- [ ] New tests are added if needed and existing tests are updated
- [ ] When applicable, versions are updated in `.cabal` and `CHANGELOG.md` files according to the
      [versioning process](https://github.com/intersectmbo/cardano-ledger/blob/master/RELEASING.md#versioning-process).
- [ ] The version bounds in `.cabal` files for all affected packages are updated. **If you change the bounds in a cabal file, that package itself must have a version increase.** (See [RELEASING.md](https://github.com/intersectmbo/cardano-ledger/blob/master/RELEASING.md#versioning-process))
- [ ] All visible changes are prepended to the latest section of a `CHANGELOG.md` for the affected packages. **New section is never added with the code changes.** (See [RELEASING.md](https://github.com/intersectmbo/cardano-ledger/blob/master/RELEASING.md#changelogmd))
- [x] Code is formatted with [`fourmolu`](https://github.com/fourmolu/fourmolu) (use `scripts/fourmolize.sh`)
- [x] Cabal files are formatted (use `scripts/cabal-format.sh`)
- [x] [`hie.yaml`](https://github.com/intersectmbo/cardano-ledger/blob/master/hie.yaml) has been updated (use `scripts/gen-hie.sh`)
- [x] Self-reviewed the diff

### CODE REVIEW

File: .github/workflows/haskell.yml [hunk 1]
```diff
@@ -46,9 +51,12 @@ jobs:
         os: [ubuntu-latest]
 
     steps:
-    - uses: actions/checkout@v3
+    - uses: actions/checkout@v4
+
     - name: Set NIGHTLY environment variable if the job was triggered by the scheduler
-      if: ${{ github.event_name == 'schedule' }}
+      if: "${{ github.event_name == 'schedule'
+            || contains(github.event.pull_request.title, 'nightly')
```
Comment (neilmayhew) [inline .github/workflows/haskell.yml:58]: This line is just to enable me to test that the fixes work. I'll take it out before we merge.

In the future, we'll probably use the manual triggering for this, but using it for this PR would mean I have to build every iteration twice.

**Review by teodanciu — APPROVED** [review]
Looks good to me.
In case it's easier to test: I have used a fork repo to play with the build and actions, to keep the things needed just for testing separate from what I wanted to merge (just in case the idea is useful to you too)

File: .github/workflows/haskell.yml (hunk 1 above)
Comment (lehins) [reply .github/workflows/haskell.yml:58]: > I'll take it out before we merge.

We don't have to take it out. I think it is reasonable to always run nightly tests if a PR has nightly in title.

**Review by lehins — APPROVED** [review]
LGTM

**Comment by lehins** [comment]
There are still flaky tests that are failing to discards. But, it is one step closer to nightly builds going green

**Comment by neilmayhew** [comment]
> Looks good to me. In case it's easier to test: I have used a fork repo to play with the build and actions, to keep the things needed just for testing separate from what I wanted to merge (just in case the idea is useful to you too)

Thanks! That's a great idea. It looks like I don't need it for now, but I'll definitely keep that in mind for the future.

**Comment by neilmayhew** [comment]
@lehins Would you do an admin merge, please? I can't get this to pass all of the matrix builds at the same time, since it's running the nightly tests.

Also, it takes so long to run that `master` has usually been updated by the time I can run it again. However, it should rebase cleanly on top of the changes merged to `master` today.

**Comment by lehins** [comment]
@neilmayhew Sounds good. I'll take care of merging this PR.

**Pull request merged**